sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala

```diff
@@ -200,19 +200,20 @@ class SparkSessionExtensions {
     optimizerRules += builder
   }
 
-  private[this] val dataSourceRewriteRules = mutable.Buffer.empty[RuleBuilder]
+  private[this] val postOperatorOptimizationRules = mutable.Buffer.empty[RuleBuilder]
 
-  private[sql] def buildDataSourceRewriteRules(session: SparkSession): Seq[Rule[LogicalPlan]] = {
-    dataSourceRewriteRules.map(_.apply(session)).toSeq
+  private[sql] def buildPostOperatorOptimizationRules(
+      session: SparkSession): Seq[Rule[LogicalPlan]] = {
+    postOperatorOptimizationRules.map(_.apply(session)).toSeq
   }
 
   /**
-   * Inject an optimizer `Rule` builder that rewrites data source plans into the [[SparkSession]].
+   * Inject an optimizer `Rule` builder that rewrites logical plans into the [[SparkSession]].
    * The injected rules will be executed after the operator optimization batch and before rules
    * that depend on stats.
    */
-  def injectDataSourceRewriteRule(builder: RuleBuilder): Unit = {
-    dataSourceRewriteRules += builder
+  def injectPostOperatorOptimizationRule(builder: RuleBuilder): Unit = {
```

Comment:
This isn't final.

I'll explain my current thinking here. We are going to use this batch to rewrite plans directly after operator optimization. I did not go for `planRewriteRules` due to the [comment](https://github.com/apache/spark/pull/30558#discussion_r533151935) made by @dongjoon-hyun and @viirya (please let me know if you don't feel that way anymore). I also did not go for `preCBORules` as there quite a few things that will happen before CBO like discussed in the original thread.

Reply:
To be honest, I do not like this name. Our catalyst is designed as a query compiler. There does not exist a concept in a query compiler called "Post Operator Optimization". If you talk about this extension point in a Database class, no one understands what it means from the name. 

My point is still the same. Here, we are adding the extension point for advanced users to add the rules between the rule-based optimizer (RBO) rules and the cost-based optimizer (CBO) rules.

Reply:
The batch/function/rule names in Optimizer.scala is not critical to me. They are private. We can change them whenever we need. However, this API is an external developer API. We should be very careful to name it. 

To me, either preCBORules or postRBORules are much better than postOperatorOptimizationRule.

Reply:
BTW, eventually, we should move all the statistics based rules from the optimizer to the planner

Reply:
It's unfortunate that we don't have a clear separation between RBO and CBO. There are RBO rules before and after the only CBO rule `CostBasedJoinReorder`.

I think the general idea of this batch is to allow people to inject special optimizer rules that can't be run together with the main operator optimizer batch. It's indeed a Spark specific thing, as the main operator optimizer batch will be run many times until reaching the fixed point, the new batch added here will be run only once.

It's really hard to do the naming here. To match the actual purpose and to be general, how about `Phase 2 Optimizer Rules` or `Run Once Optimizer Rules`?

Reply:
`Phase 2 Optimizer Rules` or `Run Once Optimizer Rules` does not explain the location of the batch. 

How many batches do we want to add? I am afraid we will add phase 2, 3, 4, .... This is endless. It looks very random. We should not expect the users need to understand/read the source code of our optimizer every new release. It is also fragile.

Reply:
I do share some of the points brought by @cloud-fan. There is no clear separation between CBO and RBO and we run quite some rule-based optimizations after the existing cost-based optimization rule. Also, we need to run this batch before early scan pushdown rules to capture possible rewrites, which, in turn, run before CBO. That’s why `preCBORules` does seem to capture what this rule does.

Our existing hook in the extensions API says `The injected rules will be executed during the operator optimization batch`. That was one of my motivations for the current name as we have resolution and post-hoc resolution rules.

That said, I don’t mean `postOperatorOptimizationRules` is a perfect name either. It just seems to better reflect the place in the Spark optimizer. I’ll be happy to discuss and iterate further.

Reply:
I'd be also interested to hear from @hvanhovell and other folks who commented.